### PR TITLE
Fix panic in speculative execution

### PR DESCRIPTION
**Pull request by Lorak-mmk**
This PR fixes linked bug by simplifying the logic inside the function.
The new logic always checks if the resolved task is the last one, and if so it returns, so it's not possible for the tasks to be exhausted without finishing the `select!`.

This PR also adds regression test for the issue. I verified that the test actually panics without the fix.

Fixes: https://github.com/scylladb/scylla-rust-driver/issues/1085

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] ~~I have provided docstrings for the public items that I want to introduce.~~
- [ ] ~~I have adjusted the documentation in `./docs/source/`.~~
- [ ] ~~I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Comment by Lorak-mmk** [comment]
Oh, it looks like passing only completed futures to select! without complete branch causes panic instead of deadlock - in that case it may be possible to write a test.

**Comment by wprzytula** [comment]
> Oh, it looks like passing only completed futures to select! without complete branch causes panic instead of deadlock - in that case it may be possible to write a test.

Good that they designed `select!` this way. Deadlocks are sad and much harder to debug.

**Comment by Lorak-mmk** [comment]
I renamed the test, because the bug is a panic not a deadlock. PR is ready for review.

File: scylla/tests/integration/retries.rs [hunk 1]
```diff
@@ -164,3 +164,70 @@ async fn retries_occur() {
...
+        );
+
+        running_proxy.running_nodes[0]
+            .change_request_rules(Some(vec![drop_connection_on_execute.clone()]));
+        running_proxy.running_nodes[1]
+            .change_request_rules(Some(vec![drop_connection_on_execute.clone()]));
+        running_proxy.running_nodes[2]
+            .change_request_rules(Some(vec![drop_connection_on_execute.clone()]));
+
+        let _result = session.execute_unpaged(&q, (2,)).await;
```
Comment (muzarski) [inline scylla/tests/integration/retries.rs:222]: Does it make sense to assert that we received an error (i.e. `unwrap_err()`)?

**Review by muzarski — APPROVED** [review]

File: scylla/tests/integration/retries.rs [hunk 2]
```diff
@@ -164,3 +164,70 @@ async fn retries_occur() {
...
+        session
+            .query_unpaged("CREATE TABLE t (a int primary key)", &[])
+            .await
+            .unwrap();
+
+        let mut q = session
+            .prepare("INSERT INTO t (a) VALUES (?)")
+            .await
+            .unwrap();
+        q.set_is_idempotent(true); // this is to allow retry to fire
```
Comment (wprzytula) [inline scylla/tests/integration/retries.rs:205]: `retry` -> `speculative execution`

File: scylla/tests/integration/retries.rs [hunk 3]
```diff
@@ -164,3 +164,70 @@ async fn retries_occur() {
...
+
+    setup_tracing();
+    let test = |proxy_uris: [String; 3], translation_map, mut running_proxy: RunningProxy| async move {
+        let se = SimpleSpeculativeExecutionPolicy {
+            max_retry_count: 2,
+            retry_interval: Duration::from_millis(1),
+        };
+        let profile = ExecutionProfile::builder()
+            .speculative_execution_policy(Some(Arc::new(se)))
+            .retry_policy(Box::new(FallthroughRetryPolicy));
```
Comment (wprzytula) [inline scylla/tests/integration/retries.rs:183]: Technically, this is a profile builder, not a profile. Maybe let's `build()` here, not later?

File: scylla/tests/integration/retries.rs (hunk 1 above)
Comment (wprzytula) [reply scylla/tests/integration/retries.rs:222]: Yes, because otherwise this test gives us weaker guarantees about correctness of underlying mechanisms.

**Review by wprzytula — CHANGES_REQUESTED** [review]

**Comment by Lorak-mmk** [comment]
Addressed review comments

**Review by wprzytula — APPROVED** [review]
Awesome that you noticed the bug! Congrats!

**Pull request merged**